src/v/net/connection.cc

```diff
@@ -95,6 +95,17 @@ std::optional<ss::sstring> is_disconnect_exception(std::exception_ptr e) {
     return std::nullopt;
 }
 
+bool is_gate_closed_exception(const std::exception_ptr& e) {
+    try {
+        rethrow_exception(e);
+    } catch (ss::gate_closed_exception&) {
+        return true;
+    } catch (...) {
+        return false;
+    }
+    __builtin_unreachable();
```

Comment:
1. this can be local scope to the file (e.g. server.cc)?
2. is the compiler really complaining about the void return path? if so, can you add `vassert(e)` at the beginning of the function to make the compiler understand, because IIRC rethrow_exception is undefined if e is nullptr anyway.

Reply:
> this can be local scope to the file (e.g. server.cc)?

That's how I started but then I saw other similar methods defined here, so moved it here. Moving it back, I prefer it to be server.cc as well.

> is the compiler really complaining about the void return path? if so, can you add vassert(e) at the beginning of the function to make the compiler understand, because IIRC rethrow_exception is undefined if e is nullptr anyway.

it complains (correctly) if there is a null check on the etpr without return. https://godbolt.org/z/z8eWPjTeh

Here I was passing a const reference because clang-tidy (mistakenly) complains about a copy though the ptr is trivially copyable. Looks like the correct way per [cppreference](https://en.cppreference.com/w/cpp/error/rethrow_exception) is to pass by copy and do a null check.